csharp/src/Ice/OutgoingAsync.cs

```diff
@@ -274,6 +291,34 @@ protected virtual bool ExceptionImpl(Ice.Exception ex)
                 return invoke;
             }
         }
+
+        protected virtual bool ExceptionImpl(Ice.RemoteException ex)
+        {
+            lock (this)
+            {
+                _ex = ex;
+                if (ChildObserver != null)
+                {
+                    ChildObserver.Failed(ex.TypeId);
+                    ChildObserver.Detach();
+                    ChildObserver = null;
+                }
+                _cancellationHandler = null;
+
+                if (Observer != null)
+                {
+                    Observer.Failed(ex.TypeId);
+                }
+                bool invoke = _completionCallback.HandleException(ex, this);
+                if (!invoke && Observer != null)
+                {
+                    Observer.Detach();
+                    Observer = null;
+                }
+                return invoke;
+            }
+        }
```

Comment:
Why duplicate this code with `Ice.Exception`? Can't we just go with a `System.Exception` parameter instead?

Reply:
The difficulty here is the TypeId (RemoteException) vs ice_id (LocalException) vs some other ID 
 (other exceptions) to give the the observers.

Reply:
Fixing.